src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,219 @@
+    // TODO: create the PositionObserver class, or create the listener for
+    // position info if inside an iframe.
+    this.element_ = element;
+  }
+
+  /**
+   * Tick function that PositionObserver can tell the class to calculate
+   * intersection.
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   */
+  tick(element, viewport) {
+    const changeEntry =
+        this.getValidIntersectionChangeEntry_(element, viewport);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   * @return {?IntersectionObserverEntry} A valid change entry.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, viewport) {
+    // Building an IntersectionObserverEntry.
+
+    let intersectionRect = element;
+    intersectionRect = rectIntersection(viewport, intersectionRect) ||
+        // No intersection.
+        layoutRectLtwh(0, 0, 0, 0);
+
+    // The element is relative to (0, 0), while the viewport moves. So, we must
+    // adjust.
+    // TODO(jridgewell, #5149): Fixed position elements must be recalculated.
+    const boundingClientRect = moveLayoutRect(element, -viewport.left,
+        -viewport.top);
+    intersectionRect = moveLayoutRect(intersectionRect, -viewport.left,
+        -viewport.top);
+
+    const ratio = intersectionRatio(intersectionRect, element);
+    const newThresholdSlot = this.getThresholdSlot_(ratio);
+    if (newThresholdSlot == this.prevThresholdSlot_) {
+      return null;
+    }
+    this.prevThresholdSlot_ = newThresholdSlot;
+
+    // Now, move the viewport to (0, 0)
+    const rootBounds = moveLayoutRect(viewport, -viewport.left, -viewport.top);
+
+    if (Date.now() == this.prevTime_) {
+      // TODO: Is this really needed?
+      return null;
+    }
+    this.prevTime_ = Date.now();
+    return /** @type {!IntersectionObserverEntry} */ ({
+      time: Date.now(),
+      rootBounds: DomRectFromLayoutRect(rootBounds),
+      boundingClientRect: DomRectFromLayoutRect(boundingClientRect),
+      intersectionRect: DomRectFromLayoutRect(intersectionRect),
+      intersectionRatio: ratio,
+    });
+  }
+
+  /**
+   * Returns the slot number that the current ratio fills in.
+   * @param {number} ratio Range from [0, 1]
+   * @return {number} Range from [0, threshold.length]
+   * @private
+   */
+  getThresholdSlot_(ratio) {
+    let startIdx = 0;
+    let endIdx = this.threshold_.length;
+    // 0 is a special case that does not fit into [small, large) range
+    if (ratio == 0) {
+      return 0;
+    }
+    let mid = Math.floor((startIdx + endIdx) / 2);
```

Comment:
All bitwise operations force 32bit ints, which just drops whatever the decimal point is. And, since `OR 0` doesn't modify any bits, it's a super quick "give me an int" operation.